### PR TITLE
Cmsis rtos v2

### PR DESCRIPTION
**Pull request by rgundi**
These patchsets introduce the CMSIS RTOS v2 API support for Zephyr.

### CODE REVIEW

**Comment by rgundi** [comment]
Shippable is failing for 2 reasons.
1) qemu_cortex_m3 - This has multiqueue scheduler enabled and multi-queue scheduler supports only a max of 32 priority levels. Cmsis_rtos_v2 has 56 priority levels. As discussed earlier with Anas, this failure can be ignored.
2) checkpatch issues - This is because of the external header file cmsis_os2.h. This can be ignored as well.

**Review by galak — DISMISSED** [review]
We need a better category for this from where we are putting the files (need to fix cmsis_v1 as well).  This would impact both the include/ and lib/.

maybe "compat_api" ?

```
include/compat_api/cmsis_v1
include/compat_api/cmsis_v2
include/compat_api/posix

lib/cmsis_rtos_v1 -> compat_api/cmsis_v1
```

Open to better names/suggestions

**Comment by rgundi** [comment]
> > 1. As discussed earlier with Anas, this failure can be ignored.
> 
> we cant ignore tests failing, this will fail for everyone after this change is merged, so please find a solution for this issue.

Well.. not sure why this will affect others. It will only affect others if they enable CMSIS RTOS v2. The fix cannot be in CMSIS as that is not desirable and will violate the basic premise of portability. The multi-queue scheduler scheduler can be doubled (to 64) which will solve this issue but i would rather let @andyross to comment on that.

**Comment by rgundi** [comment]
@galak.. open to suggestions. I have no qualms introducing another folder. Guess that blends in.

**Comment by nashif** [comment]
> Well.. not sure why this will affect others.

because this is a test that fails and will not pass CI.

**Comment by nashif** [comment]
> maybe "compat_api" ?

not compat_api please, this will be confusing, it is more about portability than compatibility (i.e. with old APIs for example)

maybe just "portable" or "portability"?

**Comment by galak** [comment]
> > maybe "compat_api" ?
> 
> not compat_api please, this will be confusing, it is more about portability than compatibility (i.e. with old APIs for example)
> 
> maybe just "portable" or "portability"?

@pfalcon usually has good ideas for such things.  @pfalcon ideas/thoughts?

**Comment by pfalcon** [comment]
Thanks. Actually, the only thing which dances in my mind regarding this is "Why Intel works on CMSIS APIs?" ;-).

> include/portable/cmsis_v1
>include/portability/cmsis_v1

IMHO, these look worse than:

> include/compat_api/cmsis_v1

The only thing which I can quickly think of is:

> include/emu_api/cmsis_v1

But that can be confusing too. But IMHO, having "api" part there is really useful. Perhaps @nashif can elaborate what exactly he doesn't like with "compat_api"?


Also have comment about:

> lib/cmsis_rtos_v1 -> compat_api/cmsis_v1

IMHO, currently the top level of Zephyr tree is pretty clean. If we push too much stuff to the top level, it may become less so. `lib/` is not overloaded currently, and IMHO good enough place for this.

**Comment by pfalcon** [comment]
Oh, and let's not forget that POSIX is also "compatibility/portability API" in a sense, so any ideas re: CMSIS should also be tested for applicability to existing POSIX stuff.

File: doc/porting/cmsis_rtos_v2.rst [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+##########################
+
+Cortex-M Software Interface Standard (CMSIS) RTOS is a vendor-independent
+hardware abstraction layer for the ARM Cortex®-M processor series and defines
+generic tool interfaces. Though it was originally defined for ARM Cortex-M
+microcontrollers alone, it could be easily extended to other microcontrollers
+making it generic. For more information on CMSIS RTOS v2, please refer
+http://www.keil.com/pack/doc/CMSIS/RTOS2/html/index.html
+
+Features not supported in Zephyr implementation:
```
Comment (dbkinder) [inline doc/porting/cmsis_rtos_v2.rst]: Needs some formatting fixes (I've laid this out as a "dictionary list", where the first line will be bold and the subsequent indented lines will be an indented paragraph):

    Features not supported in Zephyr implementation
    ***********************************************

    Kernel
       osKernelGetState, osKernelSuspend, osKernelResume are not supported.
       Also, osKernelInitialize and osKernelStart are pretty much empty.

    Threads
       osThreadJoin and osThreadDetach are not supported.

    Mutex
       osMutexPrioInherit is supported by default and is not configurable,
       we cannot select/unselect this attribute.
       osMutexRecursive is also supported by default. If this attribute is
       not set, an error is thrown when the same thread tries to acquire
       it the second time.
       osMutexRobust is not supported in Zephyr.

    Return values not supported in the Zephyr implementation
    ********************************************************

    osSemaphoreDelete
       osErrorResource (the semaphore specified by parameter
       semaphore_id is in an invalid semaphore state) is not supported.

    osMutexDelete
       osErrorResource (mutex specified by parameter mutex_id
       is in an invalid mutex state) is not supported.

    osTimerDelete
       osErrorResource (the timer specified by parameter timer_id
       is in an invalid timer state) is not supported.

    osMessageQueueReset
       osErrorResource (the message queue specified by
       parameter msgq_id is in an invalid message queue state)
       is not supported.

    osMessageQueueDelete
       osErrorResource (the message queue specified by
       parameter msgq_id is in an invalid message queue state)
       is not supported.

    osMemoryPoolFree
       osErrorResource (the memory pool specified by
       parameter mp_id is in an invalid memory pool state) is
       not supported.

    osMemoryPoolDelete
       osErrorResource (the memory pool specified by
       parameter mp_id is in an invalid memory pool state) is
       not supported.

    osEventFlagsSet, osEventFlagsClear
       osFlagsErrorUnknown (Unspecified error)
       and osFlagsErrorResource (Event flags object specified by
       parameter ef_id is not ready to be used) are not supported.

    osEventFlagsDelete
       osErrorParameter (the value of the parameter ef_id is
       incorrect) is not supported.

    osThreadFlagsSet
       osFlagsErrorUnknown (Unspecified error) and
       osFlagsErrorResource (Thread specified by parameter
       thread_id is not active to receive flags) are not supported.

    osThreadFlagsClear
       osFlagsErrorResource (Running thread is not active to
       receive flags) is not supported.

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: doc/porting/cmsis_rtos_v2.rst (hunk 1 above)
Comment (rgundi) [reply doc/porting/cmsis_rtos_v2.rst]: Done.

**Comment by rgundi** [comment]
@pfalcon @galak @nashif Any final note on the naming stuff? I personally prefer "lib/ports_api/cmsis_rtos_v2" for C files and "include/ports_api/cmsis_rtos_v2" for header files.

**Comment by rgundi** [comment]
@andyross The shippable failure is due to the multi-queue scheduler supporting a max of only 32 priority levels which is less than the total priority levels supported for cmsis_rtos_v2 (which is 56). Could you please look into this and see if that can be doubled to 64?

**Review by dbkinder — APPROVED** [review]
+1 for docs

**Comment by nashif** [comment]
> Oh, and let's not forget that POSIX is also "compatibility/portability API" in a sense, so any ideas re: CMSIS should also be tested for applicability to existing POSIX stuff.

I am not sure what you mean here. Those are 2 different interfaces that sit on top of the native zephyr kernel APIs, stacking them on top of each other is in scope here.

**Comment by nashif** [comment]
> @pfalcon @galak @nashif Any final note on the naming stuff? I personally prefer "lib/ports_api/cmsis_rtos_v2" for C files and "include/ports_api/cmsis_rtos_v2" for header files.

Maybe...

- interface/adaption
  - posix
  - cmsis_v1
  - cmsis_v2

Please address any remaining issues beside the naming of the directory, we can eventually get more input in the TSC.

**Comment by rgundi** [comment]
@nashif addressed all review comments so far. Waiting for @andyross review especially regarding shippable failure for qemu_cortex_m3 which is due to the multi-queue scheduler.

**Comment by rgundi** [comment]
@nashif... Addressed the comments. The shippable is passing.

**Review by nashif — APPROVED** [review]
@rgundi I do not see a single sample using this or the v1 APIs, we do need samples, for example implementing simple synchronization using this API. This was part of the acceptance criteria.

**Review by nashif — CHANGES_REQUESTED** [review]
@rgundi I do not see a single sample using this or the v1 APIs, we do need samples, for example implementing simple synchronization using this API. This was part of the acceptance criteria.

**Comment by rgundi** [comment]
@nashif It escaped my notice but your request does make sense. Can that be done in a different PR? That PR can address both v1 and v2.

**Review by dbkinder — APPROVED** [review]
doc change comments from #11648 included here, so LGTM

File: lib/cmsis_rtos_v2/kernel.c [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+	return (osOK);
+}
+
+/**
+ * @brief Lock the RTOS Kernel scheduler.
+ */
+int32_t osKernelLock(void)
+{
+	int temp = _current->base.sched_locked;
```
Comment (spoorthik) [inline lib/cmsis_rtos_v2/kernel.c]: This API cannot be called from ISR context. I think we need a check something like this here:

```
if (k_is_in_isr()) {
        return osErrorISR;
    }
```

**Review by spoorthik — COMMENTED** [review]
osKernelLock(), osKernelUnlock(), osKernelRestoreLock() APIs cannot be called from ISR context. So necessary check needs to be done. Please have a look.

**Review by nashif — APPROVED** [review]
naming of folder should not block this any further, we can rename where this is located along side v1 and posix when we have a better location and we get to the point of header directory structure cleanup

File: lib/cmsis_rtos_v2/kernel.c (hunk 2 above)
Comment (rgundi) [reply lib/cmsis_rtos_v2/kernel.c]: addressed. Thanks for pointing out.

**Review by spoorthik — APPROVED** [review]

**Comment by rgundi** [comment]
If there are any checkpatch errors, they are due to the external file cmsis_os2.h. Please ignore them.

**Pull request merged**